File: java/test/behaviour/util/FunctorEncoder.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.typedb.driver.test.behaviour.util;

import com.typedb.driver.api.analyze.Conjunction;
import com.typedb.driver.api.analyze.Constraint;
import com.typedb.driver.api.analyze.ConstraintVertex;
import com.typedb.driver.api.analyze.Fetch;
import com.typedb.driver.api.analyze.Function;
import com.typedb.driver.api.analyze.Pipeline;
import com.typedb.driver.api.analyze.PipelineStage;
import com.typedb.driver.api.analyze.Reducer;
import com.typedb.driver.api.analyze.VariableAnnotations;
import com.typedb.driver.api.concept.Concept;

import com.typedb.driver.jni.ConstraintExactness;
import com.typedb.driver.jni.ReturnOperationVariant;

import java.util.Arrays;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class FunctorEncoder {

    protected final Pipeline pipeline;

    public FunctorEncoder(Pipeline pipeline) {
        this.pipeline = pipeline;
    }

    private static String encodeList(Stream<String> elements) {
        return String.format("[%s]", elements.collect(Collectors.joining(",")));
    }

    private static String makeFunctor(String functor, String... encodedArgs) {
        return String.format("%s(%s)", functor, String.join(",", encodedArgs));
    }

    public static String normalizeForCompare(String functor) {
        return functor.toLowerCase().replaceAll("\\s+", "");
    }

    // The few common ones
    public String encode(com.typedb.driver.jni.Variable variable) {
        return "$" + pipeline.getVariableName(variable).orElse("_");
    }

    public static class StructureEncoder extends FunctorEncoder {
        public StructureEncoder(Pipeline pipeline) {
            super(pipeline);
        }

        private String constraintFunctor(String variant, ConstraintExactness exactness, ConstraintVertex... args) {
            String variantWithExactness = exactness == ConstraintExactness.Exact ? variant + "Exact" : variant;
            return makeFunctor(variantWithExactness, Arrays.stream(args).map(this::encode).toArray(String[]::new));
        }

        private String encode(PipelineStage stage) {
            switch (stage.getVariant()) {
                case Match:
                    return makeFunctor("Match", encode(pipeline.conjunction(stage.asMatch().block()).get()));
                case Insert:
                    return makeFunctor("Insert", encode(pipeline.conjunction(stage.asInsert().block()).get()));

                case Put:
                    return makeFunctor("Put", encode(pipeline.conjunction(stage.asPut().block()).get()));

                case Update:
                    return makeFunctor("Update", encode(pipeline.conjunction(stage.asUpdate().block()).get()));
                case Delete:
                    return makeFunctor(
                            "Delete",
                            encodeList(stage.asDelete().deletedVariables().map(this::encode)),
                            encode(pipeline.conjunction(stage.asDelete().block()).get())
                    );

                case Select:
                    return makeFunctor("Select", encodeList(stage.asSelect().variables().map(this::encode)));
                case Sort:
                    return makeFunctor("Sort", encodeList(stage.asSort().variables().map(this::encode)));
                case Require:
                    return makeFunctor("Require", encodeList(stage.asRequire().variables().map(this::encode)));
                case Offset:
                    return makeFunctor("Offset", Long.toString(stage.asOffset().offset()));
                case Limit:
                    return makeFunctor("Limit", Long.toString(stage.asLimit().limit()));
                case Distinct:
                    return makeFunctor("Distinct");
                case Reduce:
                    return makeFunctor("Reduce", encodeList(stage.asReduce().reducerAssignments().map(this::encode)), encodeList(stage.asReduce().groupBy().map(this::encode)));
                default:
                    throw new IllegalArgumentException("Unhandled stage variant", null);
            }
        }

        private String encode(PipelineStage.ReduceStage.ReduceAssignment reduceAssignment) {
            return makeFunctor("ReduceAssign", encode(reduceAssignment.assigned()), encode(reduceAssignment.reducer()));
        }

        private String encode(PipelineStage.SortStage.SortVariable sortVariable) {
            String order = sortVariable.order() == com.typedb.driver.jni.SortOrder.Ascending ? "Asc" : "Desc";
            return makeFunctor(order, encode(sortVariable.variable()));
        }

        public String encode(ConstraintVertex vertex) {
            if (vertex.isVariable()) {
                return encode(vertex.asVariable());
            } else if (vertex.isLabel()) {
                return vertex.asLabel().getLabel();
            } else if (vertex.isValue()) {
                if (vertex.asValue().isString()) {
                    return "\"" + vertex.asValue().toString() + "\"";
                } else {
                    return vertex.asValue().toString();
                }
            } else if (vertex.isNamedRole()) {
                return vertex.asNamedRoleGetName();
            } else {
                throw new IllegalArgumentException("Unexpected ConstraintVertex variant");
            }
        }

        public String encode(Constraint constraint) {
            switch (constraint.variant()) {
                case Isa: {
                    Constraint.Isa isa = constraint.asIsa();
                    return constraintFunctor("Isa", isa.exactness(), isa.instance(), isa.type());
                }
                case Has: {
                    Constraint.Has has = constraint.asHas();
                    return constraintFunctor("Has", has.exactness(), has.owner(), has.attribute());
                }
                case Links: {
                    Constraint.Links links = constraint.asLinks();
                    return constraintFunctor("Links", links.exactness(), links.relation(), links.player(), links.role());
                }
                case Sub: {
                    Constraint.Sub sub = constraint.asSub();
                    return constraintFunctor("Sub", sub.exactness(), sub.subtype(), sub.supertype());
                }
                case Owns: {
                    Constraint.Owns owns = constraint.asOwns();
                    return constraintFunctor("Owns", owns.exactness(), owns.owner(), owns.attribute());
                }
                case Relates: {
                    Constraint.Relates relates = constraint.asRelates();
                    return constraintFunctor("Relates", relates.exactness(), relates.relation(), relates.role());
                }
                case Plays: {
                    Constraint.Plays plays = constraint.asPlays();
                    return constraintFunctor("Plays", plays.exactness(), plays.player(), plays.role());
                }
                case FunctionCall: {
                    Constraint.FunctionCall functionCall = constraint.asFunctionCall();
                    return makeFunctor(
                            "FunctionCall",
                            functionCall.name(),
                            encodeList(functionCall.assigned().map(this::encode)),
                            encodeList(functionCall.arguments().map(this::encode))
                    );
                }
                case Expression: {
                    Constraint.Expression expression = constraint.asExpression();
                    return makeFunctor(
                            "Expression",
                            expression.text(),
                            this.encode(expression.assigned()),
                            encodeList(expression.arguments().map(this::encode))
                    );
                }
                case Is: {
                    Constraint.Is is = constraint.asIs();
                    // Just pass Subtypes.
                    return makeFunctor("Is", encode(is.lhs()), encode(is.rhs()));
                }
                case Iid: {
                    Constraint.Iid iid = constraint.asIid();
                    return makeFunctor("Iid", encode(iid.variable()), iid.iid());
                }
                case Comparison: {
                    Constraint.Comparison comparison = constraint.asComparison();
                    return makeFunctor(
                            "Comparison",
                            encode(comparison.lhs()),
                            encode(comparison.rhs()),
                            Constraint.Comparison.comparatorName(comparison.comparator())
                    );
                }
                case KindOf: {
                    Constraint.Kind kind = constraint.asKindOf();
                    return makeFunctor("Kind", kind.kind().toString(), encode(kind.type()));
                }
                case Label: {
                    Constraint.Label label = constraint.asLabel();
                    return makeFunctor("Label", encode(label.variable()), label.label());
                }
                case Value: {
                    Constraint.Value value = constraint.asValue();
                    return makeFunctor("Value",
                            encode(value.attributeType()),
                            value.valueType()
                    );
                }
                case Or: {
                    Constraint.Or or = constraint.asOr();
                    Stream<String> branchesEncoded = or.branches().map(conjunctionID -> {
                        Conjunction conjunction = pipeline.conjunction(conjunctionID).get();
                        return encode(conjunction);
                    });
                    return makeFunctor("Or", encodeList(branchesEncoded));
                }
                case Not: {
                    Constraint.Not not = constraint.asNot();
                    Conjunction conjunction = pipeline.conjunction(not.conjunction()).get();
                    return makeFunctor("Not", encode(conjunction));
                }
                case Try: {
                    Constraint.Try try_ = constraint.asTry();
                    Conjunction conjunction = pipeline.conjunction(try_.conjunction()).get();
                    return makeFunctor("Try", encode(conjunction));
                }
                default:
                    throw new IllegalArgumentException("Unhandled constraint variant: " + constraint.variant(), null);
            }
        }

        private String encode(Conjunction conjunction) {
            return encodeList(conjunction.constraints().map(this::encode));
        }

        public String encode(Pipeline pipeline) {
            return makeFunctor("Pipeline", encodeList(pipeline.stages().map(this::encode)));
        }

        public String encode(Function func) {
            return makeFunctor(
                    "Function",
                    encodeList(func.argument_variables().map(this::encode)),
                    encode(func.return_operation()),
                    encode(func.body())
            );
        }

        private String encode(Function.ReturnOperation returnOperation) {
            switch (returnOperation.variant()) {
                case StreamReturn:
                    return makeFunctor("Stream", encodeList(returnOperation.asStream().variables().map(this::encode)));
                case SingleReturn:
                    return makeFunctor("Single", returnOperation.asSingle().selector(), encodeList(returnOperation.asSingle().variables().map(this::encode)));
                case CheckReturn:
                    return makeFunctor("Check");
                case ReduceReturn:
                    return makeFunctor("Single", encodeList(returnOperation.asReduce().reducers().map(this::encode)));
                default:
                    throw new IllegalArgumentException("Unhandled return operation variant", null);
            }
        }

        private String encode(Reducer reducer) {
            return makeFunctor("Reducer", reducer.name(), encodeList(reducer.arguments().map(this::encode)));
        }
    }

    public static class AnnotationsEncoder extends FunctorEncoder {

        public AnnotationsEncoder(Pipeline pipeline) {
            super(pipeline);
        }

        public String encode(Pipeline pipeline) {
            return makeFunctor("Pipeline", encodeList(pipeline.stages().map(this::encode)));
        }

        public String encode(PipelineStage stage) {
            switch (stage.getVariant()) {
                case Match:
                    return makeFunctor("Match", encode(pipeline.conjunction(stage.asMatch().block()).get()));
                case Insert:
                    return makeFunctor("Insert", encode(pipeline.conjunction(stage.asInsert().block()).get()));

                case Put:
                    return makeFunctor("Put", encode(pipeline.conjunction(stage.asPut().block()).get()));

                case Update:
                    return makeFunctor("Update", encode(pipeline.conjunction(stage.asUpdate().block()).get()));
                case Delete:
                    return makeFunctor("Delete", encode(pipeline.conjunction(stage.asDelete().block()).get()));
                case Select:
                    return makeFunctor("Select");
                case Sort:
                    return makeFunctor("Sort");
                case Require:
                    return makeFunctor("Require");
                case Offset:
                    return makeFunctor("Offset");
                case Limit:
                    return makeFunctor("Limit");
                case Distinct:
                    return makeFunctor("Distinct");
                case Reduce:
                    return makeFunctor("Reduce");
                default:
                    throw new IllegalArgumentException("Unhandled stage variant", null);
            }
        }

        private String encode(VariableAnnotations variableAnnotations) {
            switch (variableAnnotations.variant()) {
                case InstanceAnnotations:
                    return FunctorEncoder.makeFunctor("Instance", FunctorEncoder.encodeList(variableAnnotations.asInstance().map(Concept::getLabel)));
                case TypeAnnotations:
                    return FunctorEncoder.makeFunctor("Type", FunctorEncoder.encodeList(variableAnnotations.asType().map(Concept::getLabel)));
                case ValueAnnotations:
                    return FunctorEncoder.makeFunctor("Value", FunctorEncoder.encodeList(variableAnnotations.asValue()));
                default:
                    throw new IllegalArgumentException("Unhandled return operation variant", null);
            }
        }

        private String encode(Conjunction conjunction) {
            Stream<String> trunkAnnotations = conjunction.annotated_variables()
                    .map(v -> this.encode(v) + ":" + encode(conjunction.variable_annotations(v)));
            String trunk = "{" + trunkAnnotations.sorted().collect(Collectors.joining(",")) + "}";
            Stream<String> branches = conjunction.constraints().map(constraint -> {
                switch (constraint.variant()) {
                    case Or:
                        return FunctorEncoder.makeFunctor(
                                "Or", FunctorEncoder.encodeList(constraint.asOr().branches().map(b -> encode(pipeline.conjunction(b).get())))
                        );
                    case Not:
                        return FunctorEncoder.makeFunctor("Not", encode(pipeline.conjunction(constraint.asNot().conjunction()).get()));
                    case Try:
                        return FunctorEncoder.makeFunctor("Try", encode(pipeline.conjunction(constraint.asNot().conjunction()).get()));
                    default:
                        return null;
                }
            }).filter(Objects::nonNull);
            return FunctorEncoder.makeFunctor("And", trunk, FunctorEncoder.encodeList(branches));
        }

        public String encode(Function func) {
            String returnStreamOrSingle = func.return_operation().variant() == ReturnOperationVariant.StreamReturn ? "Stream" : "Single";
            return FunctorEncoder.makeFunctor("Function",
                    FunctorEncoder.encodeList(func.argument_annotations().map(this::encode)),
                    FunctorEncoder.makeFunctor(returnStreamOrSingle, FunctorEncoder.encodeList(func.return_annotations().map(this::encode))),
                    encode(func.body())
            );
        }

        public String encode(Fetch fetch) {
            switch (fetch.variant()) {
                case Leaf:
                    return FunctorEncoder.encodeList(fetch.asLeaf().annotations());
                case List:
                    return FunctorEncoder.makeFunctor("List", encode(fetch.asList().element()));
                case Object:
                    Stream<String> fields = fetch.asObject().keys().map(field -> field + ":" + encode(fetch.asObject().get(field)));
                    return "{" + fields.sorted().collect(Collectors.joining(",")) + "}";
                default:
                    throw new IllegalArgumentException("Unhandled Fetch variant");
            }
        }
    }
}

File: csharp/Test/Behaviour/Util/FunctorEncoder.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

using TypeDB.Driver.Api;
using TypeDB.Driver.Api.Analyze;

using Pinvoke = TypeDB.Driver.Pinvoke;

namespace TypeDB.Driver.Test.Behaviour
{
    /// <summary>
    /// Encodes analyzed query structures into a functor-based string representation for comparison in tests.
    /// </summary>
    public abstract class FunctorEncoder
    {
        protected readonly IPipeline Pipeline;

        protected FunctorEncoder(IPipeline pipeline)
        {
            Pipeline = pipeline;
        }

        private static string EncodeList(IEnumerable<string> elements)
        {
            return $"[{string.Join(",", elements)}]";
        }

        private static string MakeFunctor(string functor, params string[] encodedArgs)
        {
            return $"{functor}({string.Join(",", encodedArgs)})";
        }

        public static string NormalizeForCompare(string functor)
        {
            return Regex.Replace(functor.ToLower(), @"\s+", "");
        }

        // The common variable encoding
        public string Encode(IVariable variable)
        {
            var name = Pipeline.GetVariableName(variable);
            return "$" + (name ?? "_");
        }

        /// <summary>
        /// Encodes the structure of analyzed queries.
        /// </summary>
        public class StructureEncoder : FunctorEncoder
        {
            public StructureEncoder(IPipeline pipeline) : base(pipeline) { }

            private string ConstraintFunctor(string variant, Pinvoke.ConstraintExactness exactness, params IConstraintVertex[] args)
            {
                string variantWithExactness = exactness == Pinvoke.ConstraintExactness.Exact ? variant + "Exact" : variant;
                return MakeFunctor(variantWithExactness, args.Select(Encode).ToArray());
            }

            private string Encode(IPipelineStage stage)
            {
                if (stage.IsMatch)
                {
                    var conjunction = Pipeline.GetConjunction(stage.AsMatch().Block);
                    return MakeFunctor("Match", Encode(conjunction!));
                }
                else if (stage.IsInsert)
                {
                    var conjunction = Pipeline.GetConjunction(stage.AsInsert().Block);
                    return MakeFunctor("Insert", Encode(conjunction!));
                }
                else if (stage.IsPut)
                {
                    var conjunction = Pipeline.GetConjunction(stage.AsPut().Block);
                    return MakeFunctor("Put", Encode(conjunction!));
                }
                else if (stage.IsUpdate)
                {
                    var conjunction = Pipeline.GetConjunction(stage.AsUpdate().Block);
                    return MakeFunctor("Update", Encode(conjunction!));
                }
                else if (stage.IsDelete)
                {
                    var deleteStage = stage.AsDelete();
                    var conjunction = Pipeline.GetConjunction(deleteStage.Block);
                    return MakeFunctor("Delete",
                        EncodeList(deleteStage.DeletedVariables.Select(Encode)),
                        Encode(conjunction!));
                }
                else if (stage.IsSelect)
                {
                    return MakeFunctor("Select", EncodeList(stage.AsSelect().Variables.Select(Encode)));
                }
                else if (stage.IsSort)
                {
                    return MakeFunctor("Sort", EncodeList(stage.AsSort().Variables.Select(Encode)));
                }
                else if (stage.IsRequire)
                {
                    return MakeFunctor("Require", EncodeList(stage.AsRequire().Variables.Select(Encode)));
                }
                else if (stage.IsOffset)
                {
                    return MakeFunctor("Offset", stage.AsOffset().Offset.ToString());
                }
                else if (stage.IsLimit)
                {
                    return MakeFunctor("Limit", stage.AsLimit().Limit.ToString());
                }
                else if (stage.IsDistinct)
                {
                    return MakeFunctor("Distinct");
                }
                else if (stage.IsReduce)
                {
                    var reduceStage = stage.AsReduce();
                    return MakeFunctor("Reduce",
                        EncodeList(reduceStage.ReducerAssignments.Select(Encode)),
                        EncodeList(reduceStage.GroupBy.Select(Encode)));
                }
                else
                {
                    throw new ArgumentException($"Unhandled stage variant: {stage.Variant}");
                }
            }

            private string Encode(IReduceAssignment reduceAssignment)
            {
                return MakeFunctor("ReduceAssign", Encode(reduceAssignment.Assigned), Encode(reduceAssignment.Reducer));
            }

            private string Encode(IReducer reducer)
            {
                return MakeFunctor("Reducer", reducer.Name, EncodeList(reducer.Arguments.Select(Encode)));
            }

            private string Encode(ISortVariable sortVariable)
            {
                string orderStr = sortVariable.Order == Pinvoke.SortOrder.Ascending ? "Asc" : "Desc";
                return MakeFunctor(orderStr, Encode(sortVariable.Variable));
            }

            public string Encode(IConstraintVertex vertex)
            {
                if (vertex.IsVariable)
                {
                    return Encode(vertex.AsVariable());
                }
                else if (vertex.IsLabel)
                {
                    return vertex.AsLabel().GetLabel();
                }
                else if (vertex.IsValue)
                {
                    var value = vertex.AsValue();
                    if (value.IsString())
                    {
                        return "\"" + value.GetString() + "\"";
                    }
                    return value.ToString()!;
                }
                else if (vertex.IsNamedRole)
                {
                    return vertex.AsNamedRole().Name;
                }
                else
                {
                    throw new ArgumentException("Unexpected ConstraintVertex variant");
                }
            }

            public string Encode(IConstraint constraint)
            {
                if (constraint.IsIsa)
                {
                    var isa = constraint.AsIsa();
                    return ConstraintFunctor("Isa", isa.Exactness, isa.Instance, isa.Type);
                }
                else if (constraint.IsHas)
                {
                    var has = constraint.AsHas();
                    return ConstraintFunctor("Has", has.Exactness, has.Owner, has.Attribute);
                }
                else if (constraint.IsLinks)
                {
                    var links = constraint.AsLinks();
                    return ConstraintFunctor("Links", links.Exactness, links.Relation, links.Player, links.Role);
                }
                else if (constraint.IsSub)
                {
                    var sub = constraint.AsSub();
                    return ConstraintFunctor("Sub", sub.Exactness, sub.Subtype, sub.Supertype);
                }
                else if (constraint.IsOwns)
                {
                    var owns = constraint.AsOwns();
                    return ConstraintFunctor("Owns", owns.Exactness, owns.Owner, owns.Attribute);
                }
                else if (constraint.IsRelates)
                {
                    var relates = constraint.AsRelates();
                    return ConstraintFunctor("Relates", relates.Exactness, relates.Relation, relates.Role);
                }
                else if (constraint.IsPlays)
                {
                    var plays = constraint.AsPlays();
                    return ConstraintFunctor("Plays", plays.Exactness, plays.Player, plays.Role);
                }
                else if (constraint.IsFunctionCall)
                {
                    var functionCall = constraint.AsFunctionCall();
                    return MakeFunctor("FunctionCall", functionCall.Name,
                        EncodeList(functionCall.Assigned.Select(Encode)),
                        EncodeList(functionCall.Arguments.Select(Encode)));
                }
                else if (constraint.IsExpression)
                {
                    var expression = constraint.AsExpression();
                    return MakeFunctor("Expression", expression.Text, Encode(expression.Assigned),
                        EncodeList(expression.Arguments.Select(Encode)));
                }
                else if (constraint.IsIs)
                {
                    var isConstraint = constraint.AsIs();
                    return MakeFunctor("Is", Encode(isConstraint.Lhs), Encode(isConstraint.Rhs));
                }
                else if (constraint.IsIid)
                {
                    var iid = constraint.AsIid();
                    return MakeFunctor("Iid", Encode(iid.Variable), iid.Iid);
                }
                else if (constraint.IsComparison)
                {
                    var comparison = constraint.AsComparison();
                    return MakeFunctor("Comparison", Encode(comparison.Lhs), Encode(comparison.Rhs), comparison.ComparatorSymbol);
                }
                else if (constraint.IsKind)
                {
                    var kind = constraint.AsKind();
                    return MakeFunctor("Kind", kind.Kind.ToString(), Encode(kind.Type));
                }
                else if (constraint.IsLabel)
                {
                    var label = constraint.AsLabel();
                    return MakeFunctor("Label", Encode(label.Variable), label.LabelValue);
                }
                else if (constraint.IsValue)
                {
                    var value = constraint.AsValue();
                    return MakeFunctor("Value", Encode(value.AttributeType), value.ValueType);
                }
                else if (constraint.IsOr)
                {
                    var or = constraint.AsOr();
                    var branchesEncoded = or.Branches.Select(conjunctionId =>
                    {
                        var conjunction = Pipeline.GetConjunction(conjunctionId);
                        return Encode(conjunction!);
                    });
                    return MakeFunctor("Or", EncodeList(branchesEncoded));
                }
                else if (constraint.IsNot)
                {
                    var not = constraint.AsNot();
                    var conjunction = Pipeline.GetConjunction(not.Conjunction);
                    return MakeFunctor("Not", Encode(conjunction!));
                }
                else if (constraint.IsTry)
                {
                    var tryConstraint = constraint.AsTry();
                    var conjunction = Pipeline.GetConjunction(tryConstraint.Conjunction);
                    return MakeFunctor("Try", Encode(conjunction!));
                }
                else
                {
                    throw new ArgumentException($"Unhandled constraint variant: {constraint.Variant}");
                }
            }

            private string Encode(IConjunction conjunction)
            {
                return EncodeList(conjunction.Constraints.Select(Encode));
            }

            public string Encode(IPipeline pipeline)
            {
                return MakeFunctor("Pipeline", EncodeList(pipeline.Stages.Select(Encode)));
            }

            public string Encode(IFunction func)
            {
                return MakeFunctor("Function",
                    EncodeList(func.ArgumentVariables.Select(Encode)),
                    Encode(func.ReturnOperation),
                    Encode(func.Body));
            }

            private string Encode(IReturnOperation returnOp)
            {
                if (returnOp.IsStream)
                {
                    return MakeFunctor("Stream", EncodeList(returnOp.AsStream().Variables.Select(Encode)));
                }
                else if (returnOp.IsSingle)
                {
                    var single = returnOp.AsSingle();
                    return MakeFunctor("Single", single.Selector, EncodeList(single.Variables.Select(Encode)));
                }
                else if (returnOp.IsCheck)
                {
                    return MakeFunctor("Check");
                }
                else if (returnOp.IsReduce)
                {
                    return MakeFunctor("Reduce", EncodeList(returnOp.AsReduce().Reducers.Select(Encode)));
                }
                else
                {
                    throw new ArgumentException($"Unhandled return operation variant: {returnOp.Variant}");
                }
            }
        }

        /// <summary>
        /// Encodes the annotations of analyzed queries.
        /// </summary>
        public class AnnotationsEncoder : FunctorEncoder
        {
            public AnnotationsEncoder(IPipeline pipeline) : base(pipeline) { }

            public string Encode(IPipeline pipeline)
            {
                return MakeFunctor("Pipeline", EncodeList(pipeline.Stages.Select(Encode)));
            }

            public string Encode(IPipelineStage stage)
            {
                if (stage.IsMatch)
                {
                    var conjunction = Pipeline.GetConjunction(stage.AsMatch().Block);
                    return MakeFunctor("Match", Encode(conjunction!));
                }
                else if (stage.IsInsert)
                {
                    var conjunction = Pipeline.GetConjunction(stage.AsInsert().Block);
                    return MakeFunctor("Insert", Encode(conjunction!));
                }
                else if (stage.IsPut)
                {
                    var conjunction = Pipeline.GetConjunction(stage.AsPut().Block);
                    return MakeFunctor("Put", Encode(conjunction!));
                }
                else if (stage.IsUpdate)
                {
                    var conjunction = Pipeline.GetConjunction(stage.AsUpdate().Block);
                    return MakeFunctor("Update", Encode(conjunction!));
                }
                else if (stage.IsDelete)
                {
                    var conjunction = Pipeline.GetConjunction(stage.AsDelete().Block);
                    return MakeFunctor("Delete", Encode(conjunction!));
                }
                else if (stage.IsSelect)
                {
                    return MakeFunctor("Select");
                }
                else if (stage.IsSort)
                {
                    return MakeFunctor("Sort");
                }
                else if (stage.IsRequire)
                {
                    return MakeFunctor("Require");
                }
                else if (stage.IsOffset)
                {
                    return MakeFunctor("Offset");
                }
                else if (stage.IsLimit)
                {
                    return MakeFunctor("Limit");
                }
                else if (stage.IsDistinct)
                {
                    return MakeFunctor("Distinct");
                }
                else if (stage.IsReduce)
                {
                    return MakeFunctor("Reduce");
                }
                else
                {
                    throw new ArgumentException($"Unhandled stage variant: {stage.Variant}");
                }
            }

            private string Encode(IVariableAnnotations variableAnnotations)
            {
                if (variableAnnotations.IsInstance)
                {
                    return MakeFunctor("Instance", EncodeList(variableAnnotations.AsInstance().Select(t => t.GetLabel())));
                }
                else if (variableAnnotations.IsType)
                {
                    return MakeFunctor("Type", EncodeList(variableAnnotations.AsType().Select(t => t.GetLabel())));
                }
                else if (variableAnnotations.IsValue)
                {
                    return MakeFunctor("Value", EncodeList(variableAnnotations.AsValue()));
                }
                else
                {
                    throw new ArgumentException($"Unhandled variable annotations variant: {variableAnnotations.Variant}");
                }
            }

            private string Encode(IConjunction conjunction)
            {
                var trunkAnnotations = conjunction.AnnotatedVariables
                    .Select(v => base.Encode(v) + ":" + Encode(conjunction.GetVariableAnnotations(v)))
                    .OrderBy(x => x);
                string trunk = "{" + string.Join(",", trunkAnnotations) + "}";

                var branches = conjunction.Constraints
                    .Select(constraint =>
                    {
                        if (constraint.IsOr)
                        {
                            return MakeFunctor("Or", EncodeList(constraint.AsOr().Branches.Select(b =>
                            {
                                var conj = Pipeline.GetConjunction(b);
                                return Encode(conj!);
                            })));
                        }
                        else if (constraint.IsNot)
                        {
                            var conj = Pipeline.GetConjunction(constraint.AsNot().Conjunction);
                            return MakeFunctor("Not", Encode(conj!));
                        }
                        else if (constraint.IsTry)
                        {
                            var conj = Pipeline.GetConjunction(constraint.AsTry().Conjunction);
                            return MakeFunctor("Try", Encode(conj!));
                        }
                        else
                        {
                            return null;
                        }
                    })
                    .Where(x => x != null)
                    .Select(x => x!);

                return MakeFunctor("And", trunk, EncodeList(branches));
            }

            public string Encode(IFunction func)
            {
                string returnStreamOrSingle = func.ReturnOperation.Variant == Pinvoke.ReturnOperationVariant.StreamReturn ? "Stream" : "Single";
                return MakeFunctor("Function",
                    EncodeList(func.ArgumentAnnotations.Select(Encode)),
                    MakeFunctor(returnStreamOrSingle, EncodeList(func.ReturnAnnotations.Select(Encode))),
                    Encode(func.Body));
            }

            public string Encode(IFetch fetch)
            {
                if (fetch.IsLeaf)
                {
                    return EncodeList(fetch.AsLeaf().Annotations);
                }
                else if (fetch.IsList)
                {
                    return MakeFunctor("List", Encode(fetch.AsList().Element));
                }
                else if (fetch.IsObject)
                {
                    var obj = fetch.AsObject();
                    var fields = obj.Keys.Select(field => field + ":" + Encode(obj.Get(field))).OrderBy(x => x);
                    return "{" + string.Join(",", fields) + "}";
                }
                else
                {
                    throw new ArgumentException($"Unhandled Fetch variant: {fetch.Variant}");
                }
            }
        }
    }
}
